design/milestone_1_2_design_doc.md

```diff
@@ -0,0 +1,430 @@
+| 1    | Keep using only **namespace** and/or **service account** granularities. | \- Since Secrets Provider runs as a separate entity, we suggest creating a dedicated service account for Secrets Provider.<br />That way, the **service account** granularity serves as an exact match and **Job** granularity is redundant<br />- No code changes | \- Service account might be reused, which allows other apps to authenticate using Secrets Provider host<br />- Break application identity convention | Free              |
+| 2    | Add support for **Job** granularity in Conjur                | Allows specific identification of the Secrets Provider **Job**, which enhances security | \- Costly; requires code changes, verify compatibility, tests, docs<br />- Redundant if **service account** serves only the Secrets Provider | 10 days           |
+
+*Decision*: Solution #1, use **namespace** and **service account** to define the Secrets Provider host in Conjur policy. We would recommend to the customer to create a dedicated service account for Secrets Provider.
+
+#### Customer experience
+
+Deployment for the Secrets Provider will be done using Helm. That way, if the 
+
+1. Configure K8s Secrets with ***conjur-map\*** metadata defined
+
+2. Create `values.yaml` that will contain the following parameters and their defaults:
+
+   2.1. Service Account for Secrets Provider (default: `secrets-provider-account`)
+
+   2.2. Role (default: `secrets-provider-role`)
+
+   2.3. RoleBinding Name (default: `secrets-provider-role-binding`)
+
+   2.4. `K8S_SECRETS` (default: `[]`)
+
+   *When parameters are not configured by customer, their defaults are used*
+
+3. Install Helm Chart for the Secrets Provider
+
+We will supply the customer with a  `values.yaml` file where all default parameters are collected. Our `values.yaml` will resemble the following:
+
+```
+# K8s permissioning defaults
+permission:
+	serviceAccountName: secrets-provider-account
+	roleName: secrets-provider-role
+	roleBinding: secrets-provider-role-binding
+	rules:
+		resources: secrets
+		verbs: [ "get", "update" ]
+
+# Secrets Provider pod defaults
+secrets-provider:
+	image: cyberark/secrets-provider-for-k8s
+	name: cyberark-secrets-provider
+	namespace: default-namespace
+	ttlSecondsAfterFinished: 0	# How long the Job should stay alive even after finishing task. Used for logging.
+	
+# Secrets Provider environment variable defaults
+secrets-provider-environment:
+	conjurApplianceUrl: https://conjur-follower.default.svc.cluster.local
+	conjurAuthnUrl: https://conjur-follower.default.svc.cluster.local/authn-k8s/authn-default
+	conjurAuthnLogin: host/conjur/authn-k8s/authn-default/default
+	conjurAccount: default
+	conjurSslCertificate: conjur-master-ca-env
+	debug=true
+	k8sSecrets=[]
+	containerMode=init
```

Comment:
I think containerMode should be removed

Reply:
@orenbm can we do this in terms of the authn-client? Does it look for either 'side-car' or 'init'?

Reply:
the default value is `sidecar` so i don't think it's the behaviour you are looking for. In case of `init` we exit the container after the first cycle. in case of `sidecar` we continue to perform the cycle after some interval.

Reply:
however, i think that providing `init` as the value where the container is not an init container is a weird behaviour. Ideally, if any value would have worked for your purposes, what would you call it?

Reply:
Adding any value for `containerMode` would demand updating also in `authn-client` right?
In terms of naming, maybe `job` (Milestone 1) and `consistentDeployment` (Milestone 2). What did you think?

Reply:
I think we should distinguish here between mode of running of secret provider process and the parameters provided in values.yaml. We should not expose a parameter of running mode of secret provider in helm chart, because it should be calculated from the fact that it is ran as Job or Deployment, and that it is not init container in this case

Reply:
@sigalsax we will not need to perform changes in the authn-client if we add a new value (want to look at the code and see why for yourself?). However, if we choose a different value than `init` and still want it to terminate after one cycle then we will need to add some code in our own repo.

i actually don't think there's a better value for `containerMode`. However, we can find another solution for finishing the secrets-provider after the first cycle. Do you have any suggestion to which of the current env vars we can use for that?